src/fairseq2/models/wav2vec2/rnnt/beam_search_gpu.py

```diff
@@ -0,0 +1,472 @@
+        self.blank = vocab_size
+        self.step_max_symbols = step_max_symbols
+        self.length_norm = length_norm
+        self.merge_beam = merge_beam
+        self.always_merge_blank = always_merge_blank
+
+    @staticmethod
+    def idx_1d_to_2d(idx, dim2):
+        return idx // dim2, idx % dim2
+
+    @staticmethod
+    def idx_2d_to_1d(idx1, idx2, dim2):
+        return dim2 * idx1 + idx2
+
+    @staticmethod
+    def repeat_like(t, like):
+        t = t.reshape([-1] + [1] * (len(like.shape) - 1))
+        t = t.repeat([1] + list(like.shape)[1:])
+        return t
+
+    def apply_length_norm(self, scores, tokens, low_score):
+        # Compute non-blank count
+        non_blank_count = (tokens != self.blank).sum(axis=1) + 1
+        # Normalize
+        normalized_scores = scores / non_blank_count
+        normalized_scores[scores == low_score] = low_score  # Don't normalize low score
+        return normalized_scores
+
+    @torch.no_grad()
+    def forward(
+        self,
+        encoder_out,
+        encoder_lengths,
+        nbest,
+    ):
+        return self.run_beam_search(encoder_out, encoder_lengths, nbest)
+
+    @torch.no_grad()
+    def run_beam_search(self, encoder_out, encoder_lengths, nbest):
+        # Create the initial values
+        device = encoder_out.device
+        float_type = encoder_out.dtype
+        bsz = encoder_out.shape[0]
+
+        # The user specifes how many non-blank tokens to use for hypo expansions.
+        # Blank tokens will always be used to, so this will be the actual tensor
+        # size
+        if self.always_merge_blank:
+            nmerge = self.merge_beam + nbest if self.merge_beam > 0 else 2 * nbest
+        else:
+            nmerge = self.merge_beam if self.merge_beam > 0 else nbest
+
+        # minimum score for a hypo in the log domain
+        low_score = -100000
+        low_score_t = torch.tensor(low_score, dtype=float_type, device=device)
+
+        # Those are used to index into various tensors
+        # [0, ... 0 (nbest times), 1, ... 1 (nbest times), ...]
+        ex_ind_nbest = torch.tensor(
+            [[i] * nbest for i in range(bsz)], dtype=torch.int64, device=device
+        ).reshape(-1)
+        # [0, ... 0 (nmerge times), 1, ... 1 (nmerge times), ...]
+        ex_ind_nmerge = torch.tensor(
+            [[i] * nmerge for i in range(bsz)], dtype=torch.int64, device=device
+        ).reshape(-1)
+        # [[blank, ..., (nbest - 1) * blank], ... bsz times]
+        blank_multiplied = torch.full(
+            [bsz, nbest], fill_value=self.blank + 1, dtype=torch.int64, device=device
+        )
+        multipier = (
+            torch.arange(nbest, dtype=torch.int64, device=device).reshape(1, -1) + 1
+        )
+        blank_multiplied *= multipier
+        blank_multiplied -= 1
+
+        # The following are used for hashing and score merging of hypos
+        # [0, 1, ..., bsz -1]
+        ex_separator = torch.arange(bsz, dtype=torch.int64, device=device)
```

Comment:
why not go straight forward ?
```suggestion
        ex_separator = torch.arange(bsz, dtype=torch.int64, device=device).reshape(-1, 1)
```